lib/Sema/TypeCheckConcurrency.cpp

```diff
@@ -1969,6 +2152,72 @@ namespace {
             ConcurrentReferenceKind::CrossActor);
       }
 
+      case ActorIsolationRestriction::DistributedActor: {
+        /// mark for later diagnostics that we have we're in a distributed actor.
+        isDistributedActor = true;
+
+        // distributed actor isolation is more strict;
+        // we do not allow any property access, or synchronous access at all.
+
+        bool continueToCheckingLocalIsolation = false;
+        // Must reference distributed actor-isolated state on 'self'.
+        auto *selfVar = getReferencedSelf(base);
+        if (!selfVar) {
+          // invocation on not-'self', is only okey if this is a distributed func
+
+          if (auto func = dyn_cast<FuncDecl>(member)) {
+            if (!func->isDistributed()) {
+              ctx.Diags.diagnose(memberLoc, diag::distributed_actor_isolated_method);
+              // TODO: offer a fixit to add 'distributed' on the member; how to test fixits? See also https://github.com/apple/swift/pull/35930/files
+              noteIsolatedActorMember(member, context, isDistributedActor);
+              return true;
+            }
+
+            assert(func->isDistributed());
+
+            if (!func->hasAsync())
+              tryMarkImplicitlyAsync(memberLoc, memberRef, context);
+
+            tryMarkImplicitlyThrows(memberLoc, memberRef, context);
+
+            // TODO: mark implicitly async here?
+            // TODO: we don't really need to do anythign with the result, dont get it?
```

Comment:
I think you should be checking the results and emitting a diagnostic if either one of those marking attempts fail. We don't want to allow something like:

```
let f = distActor.func
```

Since `func` is only partially applied (to `distActor`) and we require fully-applied calls. The type of `distActor.func` does not carry the `async` or `throws` in its type, so only its call-sites are marked for special code emission to make the accesses work.

Reply:
Ah that's a good test case, thanks!